# Cleaning and Assessing E-commerce Transaction Data

## 1. Analysis Objective

The purpose of this data analysis project is to identify best-selling products based on sales transaction data, so that more effective marketing strategies can be developed to increase revenue.

The purpose of this practice project is to assess the cleanliness and tidiness of the dataset, clean the data based on the assessment results, and prepare the dataset for further analysis.

## 2. Introduction

The original dataset records all transactions of a UK-based online retail company from December 1, 2010 to December 9, 2011. It covers the company’s business data across different countries and regions. The company mainly sells gifts for various occasions, including but not limited to birthday gifts, wedding souvenirs, Christmas gifts, and more. Its customer groups mainly include wholesalers and individual consumers, with wholesalers accounting for a relatively large proportion.

The meaning of each column is as follows:

- `InvoiceNo`: Invoice number. A 6-digit number that uniquely identifies each transaction. If this code starts with the letter "C", it indicates that the transaction was cancelled.
- `StockCode`: Product code. A 5-digit code that uniquely identifies each product.
- `Description`: Product name.
- `Quantity`: The quantity of the product in the transaction.
- `InvoiceDate`: Invoice date and time. The date and time when the transaction occurred.
- `UnitPrice`: Unit price. The price of the product in British pounds (£).
- `CustomerID`: Customer number. A 5-digit number that uniquely identifies each customer.
- `Country`: Country name. The country where the customer is located.

## 3. Load Data

Import the library needed for data analysis, then use pandas `read_csv()` function to load the raw data file `e_commerce.csv` as a DataFrame named `original_data`.

In [1]:
import pandas as pd
original_data = pd.read_csv("./e_commerce.csv")
original_data

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
...,...,...,...,...,...,...,...,...
541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,12/9/2011 12:50,0.85,12680.0,France
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,12/9/2011 12:50,2.10,12680.0,France
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,12/9/2011 12:50,4.15,12680.0,France
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,12/9/2011 12:50,4.15,12680.0,France


## 4. Data Assessment

In this section, I will assess the `original_data` DataFrame created in the previous section.

The assessment focuses on data structure and data quality. Structural issues are checked based on tidy data principles: each column should represent one variable, each row should represent one observation, and each cell should contain one value. Data quality issues include missing values, duplicated records, invalid values, and other problems that may affect the analysis.

### 4.1 Data Preview and tidiness assessment

In [2]:
original_data.sample(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
210848,555325,20978,36 PENCILS TUBE SKULLS,1,6/2/2011 11:02,2.46,NaN,United Kingdom
71391,542122,22386,JUMBO BAG PINK POLKADOT,10,1/25/2011 14:36,1.95,17082.0,United Kingdom
24733,538349,22942,CHRISTMAS LIGHTS 10 SANTAS,1,12/10/2010 14:59,8.47,NaN,United Kingdom
412343,572284,22629,SPACEBOY LUNCH BOX,4,10/23/2011 12:08,1.95,16744.0,United Kingdom
167799,551005,21559,STRAWBERRY LUNCH BOX WITH CUTLERY,1,4/26/2011 9:48,4.96,NaN,United Kingdom
413184,572307,22158,3 HEARTS HANGING DECORATION RUSTIC,8,10/23/2011 15:23,2.95,16905.0,United Kingdom
312308,564346,22197,POPCORN HOLDER,21,8/24/2011 14:54,0.85,15727.0,United Kingdom
370237,569132,21892,TRADITIONAL WOODEN CATCH CUP GAME,3,9/30/2011 13:54,1.25,17757.0,United Kingdom
279408,561321,23034,DRAWER KNOB CERAMIC BLACK,6,7/26/2011 12:59,1.45,13668.0,United Kingdom
368645,568990,84879,ASSORTED COLOUR BIRD ORNAMENT,8,9/29/2011 16:42,1.69,15933.0,United Kingdom


From a random sample of 10 rows, each column represents a variable, each row represents an observation, and each cell contains a single value. Therefore, the dataset is tidy.

### 4.2 Quality assessment

In [3]:
original_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  str    
 1   StockCode    541909 non-null  str    
 2   Description  540455 non-null  str    
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  str    
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  str    
dtypes: float64(2), int64(1), str(5)
memory usage: 33.1 MB


The dataset contains 541,909 rows and 8 columns. Most columns have complete values, but `Description` and `CustomerID` contain missing values. In addition, `InvoiceDate` is currently stored as a string, and `CustomerID` is stored as a float, so these columns may need data type conversion during data cleaning.

#### 4.2.1 Missing data assessment

Missing values are found in the `Description` and `CustomerID`, so these two columns need further analysis.

##### 4.2.1.1 `Description` assessment

In [4]:
original_data[original_data["Description"].isnull()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,12/1/2010 11:52,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,12/1/2010 14:32,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,12/1/2010 14:33,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,12/1/2010 14:33,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,12/1/2010 14:34,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535322,581199,84581,NaN,-2,12/7/2011 18:26,0.0,NaN,United Kingdom
535326,581203,23406,NaN,15,12/7/2011 18:31,0.0,NaN,United Kingdom
535332,581209,21620,NaN,6,12/7/2011 18:35,0.0,NaN,United Kingdom
536981,581234,72817,NaN,27,12/8/2011 10:33,0.0,NaN,United Kingdom


From the displayed rows, records with missing `Description` appear to have a `UnitPrice` of 0.0. However, this observation is based only on the previewed rows and needs to be verified with further checks.

In [5]:
original_data[(original_data["Description"].isnull()) & (original_data["UnitPrice"] != 0)]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country


All rows with missing `Description` have a `UnitPrice` of 0.o, suggesting that these records are unlikely to represent valid product sales. Since product descriptions are required for analyzing best-selling products, rows with missing `Description` will be removed.

##### 4.2.1.2 `CustomerID` assessment

In [6]:
original_data[original_data["CustomerID"].isnull()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,12/1/2010 11:52,0.00,NaN,United Kingdom
1443,536544,21773,DECORATIVE ROSE BATHROOM BOTTLE,1,12/1/2010 14:32,2.51,NaN,United Kingdom
1444,536544,21774,DECORATIVE CATS BATHROOM BOTTLE,2,12/1/2010 14:32,2.51,NaN,United Kingdom
1445,536544,21786,POLKADOT RAIN HAT,4,12/1/2010 14:32,0.85,NaN,United Kingdom
1446,536544,21787,RAIN PONCHO RETROSPOT,2,12/1/2010 14:32,1.66,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
541536,581498,85099B,JUMBO BAG RED RETROSPOT,5,12/9/2011 10:26,4.13,NaN,United Kingdom
541537,581498,85099C,JUMBO BAG BAROQUE BLACK WHITE,4,12/9/2011 10:26,4.13,NaN,United Kingdom
541538,581498,85150,LADIES & GENTLEMEN METAL SIGN,1,12/9/2011 10:26,4.96,NaN,United Kingdom
541539,581498,85174,S/4 CACTI CANDLES,1,12/9/2011 10:26,10.79,NaN,United Kingdom


There is a many-to-many relationship between customers and products: one customer can purchase multiple products, and one product can be purchased by multiple customers. However, since this project focuses on best-selling products rather than customer behavior, missing `CustomerID` values are not directly relevant to the main analysis.

#### 4.2.2 Duplicate data assessment

In [7]:
original_data[original_data.duplicated()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
517,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,12/1/2010 11:45,1.25,17908.0,United Kingdom
527,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,12/1/2010 11:45,2.10,17908.0,United Kingdom
537,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,12/1/2010 11:45,2.95,17908.0,United Kingdom
539,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,12/1/2010 11:45,4.95,17908.0,United Kingdom
555,536412,22327,ROUND SNACK BOXES SET OF 4 SKULLS,1,12/1/2010 11:49,2.95,17920.0,United Kingdom
...,...,...,...,...,...,...,...,...
541675,581538,22068,BLACK PIRATE TREASURE CHEST,1,12/9/2011 11:34,0.39,14446.0,United Kingdom
541689,581538,23318,BOX OF 6 MINI VINTAGE CRACKERS,1,12/9/2011 11:34,2.49,14446.0,United Kingdom
541692,581538,22992,REVOLVER WOODEN RULER,1,12/9/2011 11:34,1.95,14446.0,United Kingdom
541699,581538,22694,WICKER STAR,1,12/9/2011 11:34,2.10,14446.0,United Kingdom


In [8]:
original_data.duplicated().sum()

np.int64(5268)

The dataset contains 5268 duplicated rows, which should be removed during the data cleaning process.

#### 4.2.3 Inconsistent data assessment

Since most potential inconsistencies are difficult to predict in advance, I will first focus on checking whether the values in the `Country` column are expressed consistently.

In [9]:
original_data["Country"].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


`United Kingdom` is expressed in different forms, including `United Kingdom`, `UK`, and `U.K.`. Similarly, `United States` is expressed as both `USA` and `United States`. These country names should be standardized in the data cleaning step.

#### 4.2.4 Invalid or erroneous data assessment

In [10]:
original_data.describe()

,Quantity,UnitPrice,CustomerID
count,541909.000000,541909.000000,406829.000000
mean,9.552250,4.611114,15287.690570
std,218.081158,96.759853,1713.600303
min,-80995.000000,-11062.060000,12346.000000
25%,1.000000,1.250000,13953.000000
50%,3.000000,2.080000,15152.000000
75%,10.000000,4.130000,16791.000000
max,80995.000000,38970.000000,18287.000000


The `Quantity` and `UnitPrice` columns contain negative values, which may indicate cancelled orders, returns, or invalid transaction records. Therefore, these two columns need further analysis.

##### 4.2.4.1 `Quantity` assessment

In [11]:
original_data[original_data["Quantity"] < 0]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,12/1/2010 9:41,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,12/1/2010 9:49,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,12/1/2010 10:24,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,12/1/2010 10:24,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,12/1/2010 10:24,0.29,17548.0,United Kingdom
...,...,...,...,...,...,...,...,...
540449,C581490,23144,ZINC T-LIGHT HOLDER STARS SMALL,-11,12/9/2011 9:57,0.83,14397.0,United Kingdom
541541,C581499,M,Manual,-1,12/9/2011 10:28,224.69,15498.0,United Kingdom
541715,C581568,21258,VICTORIAN SEWING BOX LARGE,-5,12/9/2011 11:57,10.95,15311.0,United Kingdom
541716,C581569,84978,HANGING HEART JAR T-LIGHT HOLDER,-1,12/9/2011 11:58,1.25,17315.0,United Kingdom


Rows with negative `Quantity` seem to be related to cancelled orders, as many of their `InvoiceNo` values start with "C". Further verification is needed.

In [12]:
original_data[(original_data["Quantity"] < 0) & (~original_data["InvoiceNo"].str.startswith("C"))]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
2406,536589,21777,NaN,-10,12/1/2010 16:50,0.0,NaN,United Kingdom
4347,536764,84952C,NaN,-38,12/2/2010 14:42,0.0,NaN,United Kingdom
7188,536996,22712,NaN,-20,12/3/2010 15:30,0.0,NaN,United Kingdom
7189,536997,22028,NaN,-20,12/3/2010 15:30,0.0,NaN,United Kingdom
7190,536998,85067,NaN,-6,12/3/2010 15:30,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535333,581210,23395,check,-26,12/7/2011 18:36,0.0,NaN,United Kingdom
535335,581212,22578,lost,-1050,12/7/2011 18:38,0.0,NaN,United Kingdom
535336,581213,22576,check,-30,12/7/2011 18:38,0.0,NaN,United Kingdom
536908,581226,23090,missing,-338,12/8/2011 9:56,0.0,NaN,United Kingdom


Rows with negative `Quantity` are not all cancelled orders. Some records have `InvoiceNo` starting with "C", while others appear to be inventory adjustment records such as missing, lost, or damaged stock. Since these records do not represent valid product sales, they should be removed before analyzing best-selling products.

##### 4.2.4.2 `UnitPrice` Assessment

In [13]:
original_data[original_data["UnitPrice"] < 0]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
299983,A563186,B,Adjust bad debt,1,8/12/2011 14:51,-11062.06,NaN,United Kingdom
299984,A563187,B,Adjust bad debt,1,8/12/2011 14:52,-11062.06,NaN,United Kingdom


Only two rows contain negative `UnitPrice` values, and both are related to `Adjust bad debt`. Since these records do not represent valid product sales, they should be removed during the data cleaning step.

## 5. Data cleaning

### 5.1 Items to Clean

- Convert the data types of `InvoiceDate` and `CustomerID`
- Remove rows with missing `Description`
- Remove duplicated rows
- Standardize the country names for `United Kingdom` and `United States`
- Remove rows with negative `Quantity`
- Remove rows with negative `UnitPrice`

### 5.2 Back Up Original Data

In [14]:
cleaned_data = original_data.copy()

Back up the original dataset to avoid losing the raw data during cleaning.

### 5.3 Data type conversion

#### 5.3.1 Conversion of `InvoiceDate`

In [15]:
cleaned_data["InvoiceDate"] = pd.to_datetime(cleaned_data["InvoiceDate"])

Convert `InvoiceDate` from a string data type to a datetime data type.

#### 5.3.2 Conversion of `CustomerID`

In [16]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].astype("Int64").astype(str)

Convert `CustomerID`from a float data type to a string data type.

Check the format of converted `CustomerID`.

In [17]:
cleaned_data["CustomerID"]

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: str

The converted `CustomerID` still included ".0", which is not reasonable in case of `CustomerID` and should be removed.

### 5.4 Remove rows with missing `Description`

In [18]:
cleaned_data.dropna(subset = ["Description"], inplace = True)

### 5.5 Remove duplicated rows

In [19]:
cleaned_data.drop_duplicates(inplace = True)

### 5.6 Standardize the country names for United Kingdom and United States

In [20]:
cleaned_data.replace(['UK', 'U.K.'], 'United Kingdom', inplace = True)
cleaned_data.replace('USA', 'United States', inplace = True)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
...,...,...,...,...,...,...,...,...
541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,2011-12-09 12:50:00,0.85,12680,France
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,2011-12-09 12:50:00,2.10,12680,France
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,2011-12-09 12:50:00,4.15,12680,France
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2011-12-09 12:50:00,4.15,12680,France


Check the number of "UK", "U.K" and "USA" after replacement.

In [21]:
print(len(cleaned_data[cleaned_data["Country"] == "UK"]))
print(len(cleaned_data[cleaned_data["Country"] == "U.K."]))
print(len(cleaned_data[cleaned_data["Country"] == "USA"]))

0
0
0


After standardization, there are no remaining records with country names written as `UK`, `U.K`, or `USA`.

#### 5.6 Remove rows with negative `Quantity`

In [26]:
cleaned_data = cleaned_data[cleaned_data["Quantity"] >= 0]

Check the number of negative `Quantity` after replacement.

In [27]:
len(cleaned_data[cleaned_data["Quantity"] < 0])

0

After standardization, there are no remaining records with negative `Quantity`.

#### 5.7 Remove rows with negative `UnitPrice`

In [29]:
cleaned_data = cleaned_data[cleaned_data["UnitPrice"] >= 0]

Check the number of negative `UnitPrice` after replacement.

In [30]:
len(cleaned_data[cleaned_data["UnitPrice"] < 0])

0

After standardization, there are no remaining records with negative `UnitPrice`.

## 6. Save cleaned data

Save the cleaned dataset as `cleaned_data.csv` in the same folder.

In [31]:
cleaned_data.to_csv("cleaned_data.csv", index = False)

Read the saved `cleaned_data.csv` file to verify that it was exported successfully.

In [32]:
pd.read_csv("./cleaned_data.csv")

C:\Users\chenh\AppData\Local\Temp\ipykernel_27064\378420524.py:1: DtypeWarning: Columns (0: InvoiceNo) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv("./cleaned_data.csv")


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
...,...,...,...,...,...,...,...,...
525455,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,2011-12-09 12:50:00,0.85,12680.0,France
525456,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,2011-12-09 12:50:00,2.10,12680.0,France
525457,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,2011-12-09 12:50:00,4.15,12680.0,France
525458,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2011-12-09 12:50:00,4.15,12680.0,France
